crates/plugins/src/manager.rs

```diff
@@ -0,0 +1,220 @@
+    /// Spin plugins repository, it fetches the latest contents of the repository and searches for
+    /// the appropriately named and versioned plugin manifest. Parses the plugin manifest to get the
+    /// appropriate source for the machine OS and architecture. Verifies the checksum of the source,
+    /// unpacks and installs it into the plugins directory.
+    pub async fn install(
+        &self,
+        plugin_manifest: &PluginManifest,
+        plugin_package: &PluginPackage,
+    ) -> Result<()> {
+        let target = plugin_package.url.to_owned();
+        let target_url = Url::parse(&target)?;
+        let temp_dir = tempdir()?;
+        let plugin_tarball_path = match target_url.scheme() {
+            URL_FILE_SCHEME => PathBuf::from(target_url.path()),
+            _ => download_plugin(&plugin_manifest.name(), &temp_dir, &target).await?,
+        };
+        verify_checksum(&plugin_tarball_path, &plugin_package.sha256)?;
+
+        self.store
+            .untar_plugin(&plugin_tarball_path, &plugin_manifest.name())?;
+
+        // Save manifest to installed plugins directory
+        self.store.add_manifest(plugin_manifest)?;
+        println!(
+            "Plugin {} was installed successfully!",
+            plugin_manifest.name()
+        );
+        Ok(())
+    }
+
+    /// Uninstalls a plugin with a given name, removing it and it's manifest from the local plugins
+    /// directory.
+    pub fn uninstall(&self, plugin_name: &str) -> Result<()> {
+        let plugin_store = self.store();
+        let manifest_file = plugin_store.installed_manifest_path(plugin_name);
+        match manifest_file.exists() {
+            // Remove the manifest and the plugin installation directory
+            true => {
+                fs::remove_file(manifest_file)?;
+                fs::remove_dir_all(plugin_store.plugin_subdirectory_path(plugin_name))?;
+                println!("Uninstalled plugin {} successfully", plugin_name);
+            }
+            false => {
+                println!(
+                    "Plugin \"{}\" isn't present, so no changes were made",
+                    plugin_name
+                );
+            }
+        }
+        Ok(())
+    }
+
+    /// Checks manifest to see if the plugin is compatible with the running version of Spin, does
+    /// not have a conflicting name with Spin internal commands, and is not a downgrade of a
+    /// currently installed plugin.
+    pub fn check_manifest(
+        &self,
+        plugin_manifest: &PluginManifest,
+        spin_version: &str,
+        allow_downgrades: bool,
+    ) -> Result<()> {
+        // Disallow installing plugins with the same name as spin internal subcommands
+        if SPIN_INTERNAL_COMMANDS
+            .iter()
+            .any(|&s| s == plugin_manifest.name())
+        {
+            bail!(
+                "Can't install a plugin with the same name ({}) as an internal plugin",
```

Comment:
I do think `internal command` rather than `internal plugin`.  If it's internal, users won't think of it as a plugin, even if we implement it that way.